kolena/_experimental/object_detection/dataset.py

```diff
@@ -0,0 +1,220 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+from collections import defaultdict
+from typing import Any
+from typing import Dict
+from typing import Literal
+from typing import Union
+
+import pandas as pd
+
+from kolena import dataset
+from kolena._experimental.object_detection.utils import compute_optimal_f1_threshold
+from kolena._experimental.object_detection.utils import compute_optimal_f1_threshold_multiclass
+from kolena._experimental.object_detection.utils import filter_inferences
+from kolena.annotation import LabeledBoundingBox
+from kolena.annotation import ScoredLabel
+from kolena.annotation import ScoredLabeledBoundingBox
+from kolena.errors import IncorrectUsageError
+from kolena.metrics import InferenceMatches
+from kolena.metrics import match_inferences
+from kolena.workflow.metrics import match_inferences_multiclass
+from kolena.workflow.metrics import MulticlassInferenceMatches
+
+
+def single_class_datapoint_metrics(bbox_matches: InferenceMatches, thresholds: float) -> Dict[str, Any]:
+    tp = [inf for _, inf in bbox_matches.matched if inf.score >= thresholds]
+    fp = [inf for inf in bbox_matches.unmatched_inf if inf.score >= thresholds]
+    fn = bbox_matches.unmatched_gt + [gt for gt, inf in bbox_matches.matched if inf.score < thresholds]
+    scores = [inf.score for inf in tp + fp]
+    return dict(
+        TP=tp,
+        FP=fp,
+        FN=fn,
+        matched_inference=[inf for _, inf in bbox_matches.matched],
+        unmatched_ground_truth=[gt for gt, _ in bbox_matches.unmatched_gt],
+        unmatched_inference=bbox_matches.unmatched_inf,
+        count_TP=len(tp),
+        count_FP=len(fp),
+        count_FN=len(fn),
+        has_TP=len(tp) > 0,
+        has_FP=len(fp) > 0,
+        has_FN=len(fn) > 0,
+        max_confidence_above_t=max(scores) if len(scores) > 0 else None,
+        min_confidence_above_t=min(scores) if len(scores) > 0 else None,
+        thresholds=thresholds,
+    )
+
+
+def multiclass_datapoint_metrics(
+    bbox_matches: MulticlassInferenceMatches,
+    thresholds: Dict[str, float],
+) -> Dict[str, Any]:
+    tp = [inf for _, inf in bbox_matches.matched if inf.score >= thresholds[inf.label]]
+    fp = [inf for inf in bbox_matches.unmatched_inf if inf.score >= thresholds[inf.label]]
+    fn = [gt for gt, _ in bbox_matches.unmatched_gt] + [
+        gt for gt, inf in bbox_matches.matched if inf.score < thresholds[inf.label]
+    ]
+    confused = [
+        ScoredLabeledBoundingBox(
+            label=inf.label,
+            score=inf.score,
+            top_left=inf.top_left,
+            bottom_right=inf.bottom_right,
+            actual_label=gt.label,
```

Comment:
I was thinking of something similar, but adding **predicted_label** and **actual_label** to the `unmatched_ground_truth` and `unmatched_inference`. Do we also need the predicted_label/actual_label for the `unmatched_inference`?

Reply:
The confusion matrix, as currently implemented, depends on threshold, so I think it doesn't make sense to put into raw results. Same argument for "predicted vs actual".